core/src/lib/thread_util.h

```diff
@@ -0,0 +1,144 @@
+/*
+   BAREOS® - Backup Archiving REcovery Open Sourced
+
+   Copyright (C) 2018-2023 Bareos GmbH & Co. KG
+
+   This program is Free Software; you can redistribute it and/or
+   modify it under the terms of version three of the GNU Affero General Public
+   License as published by the Free Software Foundation and included
+   in the file LICENSE.
+
+   This program is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301, USA.
+*/
+#ifndef BAREOS_LIB_THREAD_UTIL_H_
+#define BAREOS_LIB_THREAD_UTIL_H_
+
+#include <mutex>
+#include <shared_mutex>
+#include <optional>
+#include <condition_variable>
+
+template <typename Mutex, template <typename> typename Lock, typename T>
```

Comment:
Ill have to think about that.  I think its pretty ingrained in the c++ community that `T` is the `value_type` of container-like types (see `optional<T>`, `vector<T, Alloc>`, etc.).
I should probably move T up front at least though.

Reply:
You have a point. However, the reader cannot see if you used `T` because it is the contained type or because you were lazy naming things :)
Reordering the Parameters is probably a good idea. Maybe there are even reasonable defaults for `Mutex` and `Lock`.